### PR TITLE
Add trimmed mean

### PR DESCRIPTION
**Pull request by sviscaino**
- Add `trimmedMean` to TDigest to compute the estimated [trimmed mean](https://en.wikipedia.org/wiki/Truncated_mean) given lower and upper percentile bounds
- Add the `trimmed_mean` as a scalar Presto function

It implements the following calculation, which as far as I understand is the same as the algorithm implemented in [python's tdigest](https://github.com/CamDavidsonPilon/tdigest/blob/master/tdigest/tdigest.py#L216).
tdigest_trimmed_mean drawio (2)
Test plan:
`mvn -Dtest="TestTDigest,TestTDigestFunctions" test`

- ensured that the returned estimated trimmed mean is ±0.05σ from the actual trimmed mean in TDigest. Tested for uniform, normal and Poisson distributions, accross all lower/upper quantile combinations. Disabled most of the tests for performance.
- followed the same logic to test the Presto function

```
== RELEASE NOTES ==

General Changes
* Added TRIMMED_MEAN() for Tdigest. The prototype of the function is:

TRIMMED_MEAN(tdigest: TDIGEST, lower_quantile: DOUBLE, upper_quantile: DOUBLE) -> DOUBLE
```

### CODE REVIEW

**Comment by amellnik** [comment]
Nit: I understand what's going on the diagram, but I think putting the weights on the x axis is going to confuse some people. It might be easier to show it similar to the way that it's done in https://arxiv.org/pdf/1902.04023.pdf
Screen Shot 2022-06-13 at 1 59 06 PM

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java [hunk 1]
```diff
@@ -136,4 +136,15 @@ public static Block destructureTDigest(@SqlType("tdigest(double)") Slice input)
         blockBuilder.closeEntry();
         return TDIGEST_CENTROIDS_ROW_TYPE.getObject(blockBuilder, blockBuilder.getPositionCount() - 1);
     }
+
+    @ScalarFunction(value = "trimmed_mean", visibility = EXPERIMENTAL)
+    @Description("Returns an estimate of the mean, after trimming to keep only the given percentile bounds")
```
Comment (amellnik) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java]: Super-nit: Maybe "Returns an estimate of the mean, excluding portions of the distribution outside the provided percentile bounds"?

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java [hunk 2]
```diff
@@ -566,6 +566,54 @@ else if (activeCentroids == 1) {
         return weightedAverage(mean[n - 1], z1, max, z2);
     }
 
+    public double trimmedMean(double l, double h)
```
Comment (amellnik) [inline presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: The comments on this seem a bit sparse compared to what's included elsewhere.  Can you add a few more comments to roughly cover the different cases here?

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java [hunk 3]
```diff
@@ -566,6 +566,54 @@ else if (activeCentroids == 1) {
...
+    {
+        checkArgument(l >= 0 && l <= 1, "l should be in [0,1], got %s", l);
+        checkArgument(h >= 0 && h <= 1, "h should be in [0,1], got %s", h);
+
+        if (unmergedWeight > 0) {
+            compress();
+        }
+
+        if (activeCentroids == 0 || l >= h) {
```
Comment (amellnik) [inline presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: Should l >= h be an error instead?  Returning NaN seems somewhat counter-intuitive for that, I would assume it would either swap them or error.

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java [hunk 4]
```diff
@@ -566,6 +566,54 @@ else if (activeCentroids == 1) {
...
+        double lowIndex = l * totalWeight;
+        double highIndex = h * totalWeight;
+
+        int n = activeCentroids;
+
+        double weightSoFar = 0;
+        double sumInBounds = 0;
+        double weightInBounds = 0;
+        for (int i = 0; i < n; i++) {
+            if (weightSoFar < lowIndex && lowIndex <= weightSoFar + weight[i] && highIndex <= weightSoFar + weight[i]) {
```
Comment (amellnik) [inline presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: A few one-line comments about what the other cases here would help readability.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestTDigestFunctions.java [hunk 5]
```diff
@@ -146,17 +147,15 @@ public void testGetQuantileAtValueOutsideRange()
         }
 
         functionAssertions.assertFunction(
-                format("quantile_at_value(CAST(X'%s' AS tdigest(%s)), %s) = 1",
-                        new SqlVarbinary(tDigest.serialize().getBytes()).toString().replaceAll("\\s+", " "),
-                        DOUBLE,
+                format("quantile_at_value(%s, %s) = 1",
+                        toSqlString(tDigest),
```
Comment (amellnik) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestTDigestFunctions.java:154]: I definitely think this is a good change, but just a head's up that the maintainers might ask it to be done in a different PR.

File: presto-main/src/test/java/com/facebook/presto/tdigest/TestTDigest.java [hunk 6]
```diff
@@ -163,6 +167,24 @@ public void testNormalDistributionLowVariance()
...
+    }
+
+    @Test(enabled = false)
+    public void testTrimmedMeanNormalDistributionLowVariance()
+    {
+        testTrimmedMeanForDistribution(new NormalDistribution(1000, 1));
+    }
+
+    @Test(enabled = false)
+    public void testTrimmedMeanNormalDistributionHighVariance()
```
Comment (amellnik) [inline presto-main/src/test/java/com/facebook/presto/tdigest/TestTDigest.java:183]: Can we also add a test for a distribution that's doesn't have any strong symmetry along the x axis -- maybe a Poisson for low λ?

**Review by amellnik — CHANGES_REQUESTED** [review]

**Comment by amellnik** [comment]
Also it would be good to go ahead and make the PR description match the format mentioned in the contributing guide, similar to #15666

**Comment by amellnik** [comment]
Also this needs documentation.

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java (hunk 3 above)
Comment (sviscaino) [reply presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: I'm not sure what makes the most sense here. While l and h outside [0, 1] really doesn't make sense, l > h could potentially be interpreted as "taking the average over an empty portion of the distribution", then it would make sense to return NaN. But I can change it if you think it's more intuitive to throw an error.
Also, I'm not sure what we should return if l == h. We could either return the mean[i] where the bounds are (the first if case), or have it error out or return NaN.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java (hunk 1 above)
Comment (sviscaino) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java]: Yes that's much better wording, I've just changed "percentile" to "quantile" to be consistent with other tdigest functions.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestTDigestFunctions.java (hunk 5 above)
Comment (sviscaino) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestTDigestFunctions.java:154]: I'll leave it there for now, I'll undo it if there are complaints 😄

File: presto-main/src/test/java/com/facebook/presto/tdigest/TestTDigest.java (hunk 6 above)
Comment (sviscaino) [reply presto-main/src/test/java/com/facebook/presto/tdigest/TestTDigest.java:183]: I added a test for a poisson distribution with λ=1. I had to increase the allowed error to 0.05 sigma for it to pass the tests consistently, it seems somewhat less accurate for strongly asymmetric distributions.

**Comment by sviscaino** [comment]
> Nit: I understand what's going on the diagram, but I think putting the weights on the x axis is going to confuse some people. It might be easier to show it similar to the way that it's done in https://arxiv.org/pdf/1902.04023.pdf Screen Shot 2022-06-13 at 1 59 06 PM

I think having the axes reversed would make it harder to visualise what's going on in the function: since we have a simplified version that doesn't do any interpolation between the centres of each centroid, it's a bit different than what's presented in the article.
I dug a bit into python's implementation of trimmed_mean, it turns out the first version published in the repo did in fact do interpolation, and it was then simplified by https://github.com/CamDavidsonPilon/tdigest/pull/46.

**Comment by sviscaino** [comment]
Added documentation and edited the PR to match the correct format

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java (hunk 3 above)
Comment (amellnik) [reply presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: For l == h I think it could also just return cdf(l) (or NaN).  I think this is probably fine to submit like this for now and see what the maintainers say.

**Comment by amellnik** [comment]
I think it looks good to squash commits and open a pull request against the main repo, nice job!
